src/historywork/WriteVerifiedCheckpointHashesWork.cpp

```diff
@@ -0,0 +1,169 @@
+{
+    return mCurrCheckpoint != 0;
+}
+
+std::shared_ptr<BasicWork>
+WriteVerifiedCheckpointHashesWork::yieldMoreWork()
+{
+    if (!hasNext())
+    {
+        throw std::runtime_error("Nothing to iterate over!");
+    }
+
+    auto& hm = mApp.getHistoryManager();
+    uint32_t freq = hm.getCheckpointFrequency();
+
+    auto lclHe = mApp.getLedgerManager().getLastClosedLedgerHeader();
+    LedgerNumHashPair lcl(lclHe.header.ledgerSeq,
+                          make_optional<Hash>(lclHe.hash));
+
+    // We form an _inner_ BatchWork that downloads a bunch of checkpoints in
+    // parallel, within our outer BatchWork, because the steps in the outer
+    // batchwork are sequentialized by the ConditionalWork that feeds ledger
+    // hashes from each VerifyLedgerChainWork to the next.
+    //
+    // If we only had the outer batching (which is how we did this initially) it
+    // would not go fast enough: we'd issue 16 curls (to fill up the 16 parallel
+    // slots -- or whatever value the user sets MAX_CONCURRENT_SUBPROCESSES to)
+    // and they would finish in "some order", but the order of freeing up each
+    // parallel-execution slot to do more work would be determined by the
+    // completion order -- each slot could only free when the slots ahead of it
+    // in sequential / verification order completed. So slot 2 could only free
+    // when slot 1 finished. Slot 3 when slots 1 _and_ 2 finished. Slot 4 when
+    // slots 1, 2 _and_ 3 finished. And so on. This meant a lot of the
+    // parallelism slots wind up waiting more than we like, and effective
+    // parallelism suffers, we're not running MAX_CONCURRENT_SUBPROCESSES at
+    // once, not even close. More like 3 or 4 curls on average, which isn't
+    // enough to keep a network busy given WAN RTT limits and connection
+    // overhead and such.
+    //
+    // To work around this we issue an _inner-batch_ of 64 parallel curls at a
+    // time, with _no_ sequential order on their completion. This ensures a much
+    // higher level of saturation of the set of parallel execution slots, at
+    // least in between the outer batching-level's effective "synchronization
+    // barriers" that happen each time a verification-step happens.
+
+    uint32_t span = 64 * freq;
+
+    uint32_t last = mCurrCheckpoint;
+    uint32_t first =
+        last <= span ? 0 : hm.firstLedgerInCheckpointContaining(last - span);
+
+    LedgerRange ledgerRange = LedgerRange::inclusive(first, last);
+    CheckpointRange checkpointRange(ledgerRange, hm);
+
+    std::string checkpointStr = std::to_string(mCurrCheckpoint);
+
+    auto tmpWork = std::make_shared<TmpDirOwnerWork>(
+        mApp, "verify-checkpoint-" + checkpointStr);
+    auto getWork = std::make_shared<BatchDownloadWork>(
+        mApp, checkpointRange, HISTORY_FILE_TYPE_LEDGER, tmpWork->getTmpDir(),
+        mArchive);
+    auto prevTrusted =
+        (mPrevVerifyWork
+             ? mPrevVerifyWork->getVerifiedMinLedgerPrev()
```

Comment:
It's a dataflow variable -- a ReaderOfWriteOnceSharedVar -- that will be filled in by mPrevVerifyWork when it runs. I'll add a comment and (cough cough) an explicit type rather than an auto, for clarity :)

Reply:
oh right, the whole point of ReaderOfWriteOnceSharedVar :) ok!